Modified behavior of --help/-h and added commands -H/--help--all

Mendi03 opened the pull request:
Closes  #1453 

- Changed the behavior of `--help` / `-h`. Some subcommands and there respective help messages were removed since they are only used for development purposes, it will now only show less subcommands. 
- Added the subcommand `--help-all` / `-H`, which will now do what `--help` did before. It will show all subcommands and their help message.

Output of `-h` and `--help`:
```console
optional arguments:
  -h, --help            show this help message and exit
  -V, --version         show program's version number and exit
  -c CONFIGFILE, --config CONFIGFILE
                        Specify Path to Configuration File
  -d, --debug           Stream log messages to stdout
  -l {DEBUG,INFO,WARNING,ERROR,CRITICAL}, --loglevel {DEBUG,INFO,WARNING,ERROR,CRITICAL}
                        Filter log messages based on logging level
  --editor {vi,vim,emacs,nano}
                        Select your preferred editor when opening files.
  --view-log            Show content of last log
  --logpath             Print full path to last log file
  --print-log           Print content of last log without pagination
  --color COLOR         Print output of table with the selected color.
  --no-color            Disable colored output
  --helpcolor           Print available color options in a table format.
  -r REPORT, --report REPORT
                        Specify path to test report file
  -H, --help--all       Show help for all commands

COMMANDS:
  
    build (bd)          Build and Run test
    buildspec (bc)      Buildspec Interface
    config (cg)         Query buildtest configuration
    report (rt)         Query test report
    inspect (it)        Inspect a test based on NAME or ID
    path                Show path attributes for a given test
    history (hy)        Query build history
    schema              List schema contents and examples
    cdash               Upload test to CDASH server
    cd                  change directory to root of test given a test name
    clean               Remove all generate files from buildtest including test directory, log files, report file, buildspec cache, history files.
    debugreport (debug)
                        Display system information and additional information for debugging purposes.
    stats               Show test statistics for given test
    info                Show details regarding current buildtest setup
    show (s)            buildtest command guide

```

Output of `-H` and `--help-all`

```console
optional arguments:
  -h, --help            show this help message and exit
  -V, --version         show program's version number and exit
  -c CONFIGFILE, --config CONFIGFILE
                        Specify Path to Configuration File
  -d, --debug           Stream log messages to stdout
  -l {DEBUG,INFO,WARNING,ERROR,CRITICAL}, --loglevel {DEBUG,INFO,WARNING,ERROR,CRITICAL}
                        Filter log messages based on logging level
  --editor {vi,vim,emacs,nano}
                        Select your preferred editor when opening files.
  --view-log            Show content of last log
  --logpath             Print full path to last log file
  --print-log           Print content of last log without pagination
  --color COLOR         Print output of table with the selected color.
  --no-color            Disable colored output
  --helpcolor           Print available color options in a table format.
  -r REPORT, --report REPORT
                        Specify path to test report file
  -H, --help--all       Show help for all commands

COMMANDS:
  
    build (bd)          Build and Run test
    buildspec (bc)      Buildspec Interface
    config (cg)         Query buildtest configuration
    report (rt)         Query test report
    inspect (it)        Inspect a test based on NAME or ID
    path                Show path attributes for a given test
    history (hy)        Query build history
    schema              List schema contents and examples
    cdash               Upload test to CDASH server
    unittests (test)    Run buildtest unit tests
    stylecheck (style)  Run buildtest style checks
    cd                  change directory to root of test given a test name
    clean               Remove all generate files from buildtest including test directory, log files, report file, buildspec cache, history files.
    schemadocs          Open buildtest schema docs in browser
    docs                Open buildtest docs in browser
    debugreport (debug)
                        Display system information and additional information for debugging purposes.
    stats               Show test statistics for given test
    info                Show details regarding current buildtest setup
    show (s)            buildtest command guide
    tutorial-examples   Generate documentation examples for Buildtest Tutorial
```

Mendi03 committed:
Modified behavior of --help and added --help--all command

Modified the behavior of --help and added new command --help--all

Mendi03 committed:
Formatted code using stylecheck

Mendi03 committed:
Added some comments to help modify --help and --help--all menu in the future

shahzebsiddiqui commented on bash_completion.sh:
```diff
@@ -458,7 +458,7 @@ _buildtest ()
       local cmds="build buildspec cd cdash clean config debugreport docs help info inspect history path report schema schemadocs stats stylecheck tutorial-examples unittests"
       local alias_cmds="bd bc cg debug it h hy rt style test"
-      local opts="--color --config --debug --editor --help --helpcolor --logpath --loglevel --print-log --no-color --report --version --view-log -c -d -h -l -p -r -V"
+      local opts="--color --config --debug --editor --help --helpcolor --help--all --logpath --loglevel --print-log --no-color --report --version --view-log -c -d -h -l -p -r -V -H"
```
Please sort them so put -H before -V thats how it'd setup it helps in tab completion

Mendi03 replied:
Just did it.

Mendi03 committed:
wrote -H before -V

shahzebsiddiqui commented on buildtest/cli/__init__.py:
```diff
@@ -12,6 +13,11 @@
 
+# Variables needed to show all sub commands and their help mesaage
+help1 = "-H" in sys.argv
+help2 = "--help--all" in sys.argv
```
can you rename to `--help-all` so no double dash before `all`. Long options are double dash like `--XYZ` and short options are single dash `-X` thats how argparse refers to short vs long options. 
![Screenshot 2023-07-21 at 11 26 20 AM](https://github.com/buildtesters/buildtest/assets/12942230/fb849a58-86f9-4768-a338-a6e8053f2ebe)

Mendi03 replied:
Just finished shortening it!

Mendi03 committed:
shortened --help--all to --help-all

Mendi03 committed:
Merge branch 'buildtesters:devel' into help_all

shahzebsiddiqui commented on buildtest/cli/__init__.py:
```diff
@@ -267,7 +267,7 @@ def get_parser():
     parser.add_argument("-r", "--report", help="Specify path to test report file")
     parser.add_argument(
-        "-H", "--help--all", help="Show help for all commands", action="help"
+        "-H", "--help-all", help="Show help for all commands", action="help"
```
You can say the following help message `List all commands and options`

shahzebsiddiqui commented on bash_completion.sh:
```diff
@@ -133,7 +133,7 @@ _buildtest ()
   COMPREPLY=()   # Array variable storing the possible completions.
 
-  declare -a buildtest_opts=("--color" "--config" "--debug" "--editor" "--help" "--helpcolor" "--logpath" "--loglevel" "--print-log" "--no-color" "--report" "--version" "--view-log" "-c" "-d" "-h" "-l" "-p" "-r" "-V")
+  declare -a buildtest_opts=("--color" "--config" "--debug" "--editor" "--help" "--helpcolor" "--help-all" "--logpath" "--loglevel" "--print-log" "--no-color" "--report" "--version" "--view-log" "-c" "-d" "-h" "-l" "-p" "-r" "-V" "-H")
```
please reorder the options so `-H` first followed by `-V`

shahzebsiddiqui commented on buildtest/cli/__init__.py:
```diff
@@ -12,6 +13,11 @@
+# Variables needed to show all sub commands and their help mesaage
+help1 = "-H" in sys.argv
+help2 = "--help-all" in sys.argv
+show_all_help = help1 or help2
```
this seems silly we dont need two variables `help1` and `help2`. It also not readable.

This can be a one liner

```python
show_all_help = "-H" in sys.argv or "--help-all" in sys.argv
```

Mendi03 replied:
Ill make sure to fix that!

shahzebsiddiqui commented on buildtest/cli/__init__.py:
```diff
@@ -353,8 +362,16 @@ def misc_menu(subparsers):
+        subparsers.add_parser("docs", help="Open buildtest docs in browser")
+    else:
+        subparsers.add_parser("schemadocs")
+        subparsers.add_parser("docs")
```
can we avoid having redundant if/else statement instead just have one place perhaps at bottom where we add all the hidden parsers. 

You might just want to declare some variable for hidden parsers maybe here is one idea

```python
hidden_parser = { 'docs' : "Open buildtest docs in browser", 
                               'schemadocs': "..."
                             }
```

Then you could do the following

```python                             
for command, help_msg in hidden_parser.items():
   subparser.add_parser(command, help=help_msg)
```

If you want to capture aliases also then you will need to update the dictionary with more keys. This would require a bit refactoring.

Another question is do we need to do `subparsers.add_parser` in the `else` statement? I take it this is not required since the argparse will only need to build add hidden parsers when `-H` is specified otherwise it is not needed

Mendi03 committed:
Merge branch 'buildtesters:devel' into help_all

Mendi03 committed:
created method add_hidden_parser

wrote the method to avoid having redundant if/else statement when trying to display hidden help messages

Mendi03 committed:
removed a comma

shahzebsiddiqui commented on buildtest/cli/__init__.py:
```diff
@@ -430,9 +453,17 @@ def unittest_menu(subparsers):
-        "unittests", help="Run buildtest unit tests", aliases=["test"]
-    )
+    # Subcommands that do not need to be shown in --help
+    h_unittests = {
```
I understand what you are doing and this can be improved if we just have one dictionary where keys are the actual commands we can declare everything once and just call the method once as pose to calling it multiple times.

This will require a refactoring for method `add_hidden_parser` since dictionary lookup will be changed.

Mendi03 replied:
You're right, that would be more efficient.

shahzebsiddiqui commented on buildtest/cli/__init__.py:
```diff
@@ -1295,3 +1331,24 @@ def cdash_menu(subparsers):
+    if "command" not in hidden_parser.keys():
+        for command, help_msg in hidden_parser.items():
+            subparser.add_parser(command, help=help_msg)
+        return
```
Remember that python  will return `None` by default from method so this `return` statement is redundant

shahzebsiddiqui commented on buildtest/cli/__init__.py:
```diff
@@ -1295,3 +1331,24 @@ def cdash_menu(subparsers):
+
+
+def add_hidden_parser(subparser, hidden_parser):
+    """This method adds a help message to certain parsers
```
Change doctoring to 

This method will add parser for hidden commands that can be shown when using --help-all

shahzebsiddiqui commented on buildtest/cli/__init__.py:
```diff
@@ -1295,3 +1331,24 @@ def cdash_menu(subparsers):
+    """
+    if "command" not in hidden_parser.keys():
+        for command, help_msg in hidden_parser.items():
+            subparser.add_parser(command, help=help_msg)
```
Since you are building the parser here might as well capture the aliases too when building the parser so you don't need to call this method again.

Just have this method called once in such a way the dictionary is setup to take care of the logic.

For commands that don't have an alias then don't define it in the dictionary

Mendi03 replied:
I just made these changes, please let me know if they're correct

shahzebsiddiqui replied:
thanks

Mendi03 committed:
Updated logic for the --help-all command.

help_all method is only called once in get_parser method.

Mendi03 committed:
stylecheck fix

Mendi03 committed:
updated logic of method help_all, removed method add hidden parser

shahzebsiddiqui commented on buildtest/cli/__init__.py:
```diff
@@ -1295,3 +1303,26 @@ def cdash_menu(subparsers):
+
+
+def help_all(subparsers):
+    """This method will add parser for hidden command that can be shown when using --help-all/-H
```
please format `--help-all/-H` by surrounding with backticks "``". This will render properly in docstring

shahzebsiddiqui commented on buildtest/cli/__init__.py:
```diff
@@ -313,6 +320,9 @@ def get_parent_parser():
 
+    # Displays all hidden comands
+    if show_all_help:
+        help_all(subparsers)
```
can you call this method at end of method so that these options show at bottom instead of top. Right now i see them as follows

```console
COMMANDS:
  
    tutorial-examples   Generate documentation examples for Buildtest Tutorial
    docs                Open buildtest docs in browser
    schemadocs          Open buildtest schema docs in browser
    unittests (test)    Run buildtest unit tests
    stylecheck (style)  Run buildtest style checks
    build (bd)          Build and Run test
    buildspec (bc)      Buildspec Interface
    config (cg)         Query buildtest configuration
```

Mendi03 committed:
minor fixes

changes based on feedback

shahzebsiddiqui reviewed (approved):
@Mendi03 this looks good. Thanks for adding support for this feature!
